src/Orchard.Web/Modules/Upgrade/Controllers/ProjectionsController.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web.Mvc;
+using Orchard;
+using Orchard.ContentManagement;
+using Orchard.ContentManagement.Handlers;
+using Orchard.ContentManagement.MetaData;
+using Orchard.Data;
+using Orchard.Environment.Features;
+using Orchard.Localization;
+using Orchard.Logging;
+using Orchard.Projections.Handlers;
+using Orchard.Security;
+using Orchard.UI.Admin;
+using Orchard.UI.Notify;
+using Upgrade.ViewModels;
+
+namespace Upgrade.Controllers {
+    [Admin]
+    public class ProjectionsController : Controller {
+        private readonly IContentDefinitionManager _contentDefinitionManager;
+        private readonly IOrchardServices _orchardServices;
+        private readonly IFeatureManager _featureManager;
+        private readonly Lazy<IEnumerable<IContentHandler>> _handlers;
+        private readonly ITransactionManager _transactionManager;
+
+        public ProjectionsController(
+            IContentDefinitionManager contentDefinitionManager,
+            IOrchardServices orchardServices,
+            ITransactionManager transactionManager,
+            IFeatureManager featureManager,
+            Lazy<IEnumerable<IContentHandler>> handlers) {
+            _contentDefinitionManager = contentDefinitionManager;
+            _orchardServices = orchardServices;
+            _transactionManager = transactionManager;
+            _featureManager = featureManager;
+            _handlers = handlers;
+            Logger = NullLogger.Instance;
+        }
+
+        public Localizer T { get; set; }
+        public ILogger Logger { get; set; }
+
+        public ActionResult Index() {
+            var viewModel = new MigrateViewModel { ContentTypes = new List<ContentTypeEntry>() };
+            foreach (var contentType in _contentDefinitionManager.ListTypeDefinitions().OrderBy(c => c.Name)) {
+                // only display parts with fields
+                if (contentType.Parts.Any(x => x.PartDefinition.Fields.Any())) {
+                    viewModel.ContentTypes.Add(new ContentTypeEntry { ContentTypeName = contentType.Name });
+                }
+            }
+
+            if (!viewModel.ContentTypes.Any()) {
+                _orchardServices.Notifier.Warning(T("There are no content types with custom fields"));
+            }
+
+            if (!_featureManager.GetEnabledFeatures().Any(x => x.Id == "Orchard.Fields")) {
+                _orchardServices.Notifier.Warning(T("You need to enable Orchard.Fields in order to migrate current fields."));
+            }
+
+            return View(viewModel);
+        }
+
+        [HttpPost, ActionName("Index")]
+        public ActionResult IndexPOST() {
```

Comment:
Shouldn't it do it by batch of content items? If there are too many of them this will fail. There are some examples doing it in other upgrade controllers.

Reply:
Can you address me to an example where I can copy from? thanks.

Reply:
Found it by myself in InfosetController.cs